File: hawg.py
import sys
import json
from operator import itemgetter

ID_attributes = {
    'root': {
        '@type': '@id'
    },
    'sourceSelector': {
        '@type': '@id',
        'defaultType': 'Selector'
    },
    'member': {
        '@type': '@id'
    },
    'baseURL': {
        '@type': '@id',
        'defaultType': 'BaseURL'
    },
    'backgroundImage': {
        '@type': '@id',
        'defaultType': 'DataSource'
    },
    'dataSource': {
        '@type': '@id',
        'defaultType': 'DataSource'

    },
    'annotation': {
        '@type': '@id',
        'defaultType': 'Annotation'

    },
    'renderOption': {
        '@type': '@id',
        'defaultType': 'Style'
    }
}

def listify(t):
    return t if type(t) == list else [t]

_blank_node_counter = 0
def blank_node_id():
    global _blank_node_counter
    _blank_node_counter += 1
    return '_:b{:0>5}'.format(_blank_node_counter)


def flatten(doc):
    blank_nodes = []
    for node in doc:
        for field_name, field_val in node.items():
            if not field_name in ID_attributes:
                continue
            field_val_list = listify(field_val)
            new_field_val = set()
            for v in field_val_list:
                if type(v) == str:
                    new_field_val.add(v)
                    continue
                try:
                    v_id = v['@id']
                except KeyError:
                    v_id = v['@id'] = blank_node_id()
                
                if not '@type' in v:
                    try:
                        v['@type'] = ID_attributes[field_name]['defaultType']
                    except KeyError:
                        pass
                new_field_val.add(v_id)
                blank_nodes.append(v)
            node[field_name] = list(new_field_val)
        try:
            node['@type'] = listify(node['@type'])
        except KeyError:
            pass
    # recurse if needed
    if blank_nodes:
        blank_nodes = flatten(blank_nodes)
    return doc + blank_nodes

def index_by_id(doc):
    id_table = {}
    # build the index
    for node in flatdoc:
        id_table[node['@id']] = node
    return id_table


def dereference(id_table):
    deref_table = {}
    # change ids to in-memory references
    for node_name, node in id_table.items():
        node_copy = {}
        for pred, obj_name_list in node.items():
            if pred not in ID_attributes:
                node_copy[pred] = node[pred]
                continue

            node_copy[pred] = [id_table[objname] for objname in obj_name_list]
        deref_table[node_name] = node_copy
    return deref_table

def reverse_id_index(doc):
    # create reverse id table
    revid_table = {}          
    for node in doc:
        for pred, objlist in node.items():
            if pred not in ID_attributes:
                continue
            for o in objlist:
                try:
                    revobj = revid_table[o['@id']]
                except KeyError:
                    revobj = revid_table[o['@id']] = {
                        '@id': node['@id'],
                        '@type': node['@type']
                    }
                try:
                    revsubj = revobj[pred]
                except KeyError:
                    revsubj = revobj[pred] = []
                if node not in revsubj:
                    revsubj.append(node)
    return revid_table

def index_by_type(doc):
    type_table = {}
    for node in doc:
        for pred, value in node.items():
            if pred != '@type':
                continue
            for typename in value:
                try:
                    type_entry = type_table[typename]
                except KeyError:
                    type_entry = type_table[typename] = []
                if node not in type_entry:
                    type_entry.append(node)

    return type_table

def node_has_type(node, types):
    types = listify(types)
    try:
        node_types = node['@type']
    except KeyError:
        return False
    for nt in node_types:
        for t in types:
            if nt == t:
                return True
    return False

def get_node_tree(node, fieldname=None, traversed=None):
    if not node:
        return []
    if traversed is None:
        traversed = set()
    if not fieldname:
        fieldname = 'member'
    
    for n in listify(node):
        if n in traversed:
            continue
        traversed.add(n)
        try:
            children = node[fieldname]
        except KeyError:
            continue
        for c in children:
            get_node_tree(children, fieldname, traversed)
    return iter(traversed)


class HAWG(object):
    def __init__(self, doc=None):
        self.id_table = {}
        self.revid_table = {}
        self.type_table = {}
        if doc:
            self.parse(doc)

    def parse(self, doc):
        flattened = flatten(doc)
        self.id_table = dereference(index_by_id(flattened))
        self.revid_table = dereference(reverse_index_by_id(flattened))
        self.type_table = index_by_type(self.id_table.values())

    def nodes(self):
        for n in self.id_table.values():
            yield n

    def get_header_node(self):
        return self.id_table['#__header__']


    def create_node(self, node_type, node_id=None, ):
        if node_id is None:
            node_id = blank_node_id()

        node_type = listify(node_type)
        node = {
            '@id': node_id,
            '@type': node_type
        }
        self.id_table[node['@id']] = node
        for t in node_type:
            if t not in self.type_table:
                self.type_table[t] = [node]
            else:
                self.type_table[t].append(node)
        return node

    def get_node_by_id(self, node_id):
        return self.id_table[node_id]

    def compose_field_values(self, node_or_id, field_name,  
                       modifier_type=None, subfields=None):
        try:
            node = self.id_table[node_or_id]
        except TypeError:
            node = node_or_id
        try:
            modifier_nodes = node[field_name]
        except KeyError:
            return {}
        
        if modifier_type:
            modifier_type = listify(modifier_type)

        modifier_dict = {}
        for a in modifier_nodes:
            if modifier_type and a.get('@type', None) not in modifier_type:
                continue
            for field, value in a.items():
                if field[0] == '@':
                    continue
                if subfields is not None and field not in subfields:
                    continue
                modifier_dict[field] = value
        return modifier_dict

    def get_nodes_with_annotation(self, 
                                filt=None,
                                node_type=None, 
                                annotation_type=None,
                                field_name=None):
        if not annotation_type:
            annotation_type = 'Annotation'
        if not field_name:
            field_name = 'annotation'
        if node_type:
            node_type = listify(node_type)
        
        annot_nodes = self.type_table[annotation_type]
        annot_nodes = filter(filt, self.type_table[annotation_type])
        annot_node_ids = [x['@id'] for x in annot_nodes]
        subj_nodes = []
        for i in annot_node_ids:
            subj_nodes.extend(self.revid_table[i][field_name])

        if node_type:
            subj_nodes = [n for n in subj_nodes if node_has_type(n, node_type)]
        return subj_nodes

if __name__ == '__main__':
    with open(sys.argv[1]) as fp:
        doc = json.load(fp)

        h = HAWG(doc)
        header = h.get_header_node()
        root = header['root']

        s = h.get_nodes_with_annotation(lambda x: x.get('name', None) == 'Muscles', node_type=['Structure', 'Group'])
        print([x['@id'] for x in s])
        for root in header['root']:
            print(root['@id'], h.compose_field_values(root, 'renderOption'),
            h.compose_field_values(root, 'annotation'))

File: hawg.js
let shortid = require('shortid');
let _ = require('lodash');

let ID_attributes = {
    'root': {
        '@type': '@id'
    },
    'sourceSelector': {
        '@type': '@id',
        'defaultType': 'Selector'
    },
    'member': {
        '@type': '@id'
    },
    'baseURL': {
        '@type': '@id',
        'defaultType': 'BaseURL'
    },
    'backgroundImage': {
        '@type': '@id',
        'defaultType': 'DataSource'
    },
    'dataSource': {
        '@type': '@id',
        'defaultType': 'DataSource'

    },
    'annotation': {
        '@type': '@id',
        'defaultType': 'Annotation'

    },
    'renderOption': {
        '@type': '@id',
        'defaultType': 'Style'
    }
};

function new_node_id() {
    let sid = shortid.generate();
    return `_:b${sid}`;
}

function flatten_doc(doc) {
    return _.flatMap(doc, flatten_node);
}

function index_doc(doc) {
    let index = {};
    _.forIn(doc, node => {
        index[node['@id']] = node
    });
    return index;
}

function reverse_node_index(node_index) {
    let rev_index = {};
    _.forOwn(node_index, (node, node_name) => {
        rev_index[node_name] = {
            '@id': node['@id']
        };
        if (_.has(node, '@type')) {
            rev_index[node_name]['@type'] = node['@type']
        }
    });

    _.forOwn(node_index, node => {
        _.forOwn(node, (field_value, field_name) => {
            if(_.has(ID_attributes, field_name)) {
                _.forEach(field_value, obj_node => {

                    let rev_subj = rev_index[obj_node['@id']];
                    let rev_obj = node;
                    if(!_.has(rev_subj, field_name)) {
                        rev_subj[field_name] = [];
                    }
                    if(!_.has(rev_subj[field_name], rev_obj)) {
                        rev_subj[field_name].push(rev_obj);
                    }
                });
            }
        });
    });
    return rev_index;
}

function dereference(node_index, existing_index) {
    let ret_index = _.mapValues(node_index, node => _.clone(node));
    let ret_index_nodes = _.values(ret_index);
    _.forOwn(ret_index_nodes, node => {
        _.forOwn(node, (field_value, field_name) => {
            if (_.has(ID_attributes, field_name)) {
                let ret = _.map(field_value, x => ret_index[x]);
                node[field_name] = ret;
            }
            else {
                node[field_name] = field_value;
            }
        });
    });
    return ret_index;
}

function compose_field_values(node, field_name, opts) {
    opts = opts || {}
    let modifier_type = opts.modifier_type ? _.castArray(opts.modifier_type) : null;
    let subfields = opts.subfields || null;

    if(!_.has(node, field_name)) {
        return {}
    }
    let modifier_dict = {};
    let modifier_nodes = node[field_name];
    _.forEach(modifier_nodes, mnode => {
        if (!modifier_type || _.includes(modifier_type, _.get(mnode, '@type', null))) {
            _.forOwn(mnode, (value, field) => {
                if (field[0] !== '@' && (!subfields || _.includes(subfields, field))) {
                    modifier_dict[field] = value;
                }
            });
        }
    });
    return modifier_dict;
}

function flatten_node(node) {
    let new_node_list = [];
    let node_copy = {};

    let flattened = _.mapValues(node, (field_value, field_name) => {
        if (!_.has(ID_attributes, field_name)) {
            return field_value;
        }
        else {
            let field_value_list = _.uniq(_.castArray(field_value));
            let new_field_values = [];
            _.forEach(field_value_list, field_value => {
                if (_.isString(field_value)) {
                    // must be a reference, and that's OK
                    new_field_values.push(field_value);
                }
                else {
                    // must be an object
                    // create a new node that is a copy of old so we can modify with @id and possibly @type
                    var field_value_node = _.clone(field_value);
                    if (!_.has(field_value_node, '@id')) {
                        field_value_node['@id'] = new_node_id();
                    }
                    if (!_.has(field_value_node, '@type') && _.has(ID_attributes[field_name], 'defaultType')) {
                        field_value_node['@type'] = ID_attributes[field_name]['defaultType'];
                    }
                    new_field_values.push(field_value_node['@id']);
                    new_node_list = new_node_list.concat(flatten_node(field_value_node));
                }
            });
            return new_field_values;
        }
    });
    // make sure @type is an array
    if(_.has(flattened, '@type')) {
        flattened['@type'] = _.castArray(flattened['@type']);
    }
    new_node_list.push(flattened);
    return new_node_list;
}

function merge_indexed_docs(doc_index0, doc_index1) {
    return _.assignInWith({}, doc_index0, doc_index1, (node0, node1) => {
        if(!_.isUndefined(node0)){
            return merge_nodes(node0, node1);
        }
    });
}

function merge_nodes(n0, n1) {
    return _.assignInWith({}, n0, n1, (n0val, n1val, field_name) => {
        if (_.has(ID_attributes, field_name) && !_.isUndefined(n0val)) {
            return _.uniq(_.concat(n0val, n1val));
        }
    });
}

function clone_node(n) {
    return _.clone(n);
}


class HAWG {
    constructor(doc) {
        this.id_table = {};
        this.node_table = {};
        this.reverse_node_table = {};
        if (doc) {
            this.parse(doc)
        };
    }

    parse(doc) {
        let indexed_doc = index_doc(flatten_doc(doc));
        this.id_table = merge_indexed_docs(this.id_table, indexed_doc);
        this.node_table = dereference(this.id_table);
        this.reverse_node_table = reverse_node_index(this.node_table);
    }


    get_node_by_id(id) {
        return this.node_table[id];
    }

    get_nodes_by_id(idlist) {
        return _.map(idlist, i => this.node_table[i]);
    }

    get_node_parent(node, fieldname) {
        fieldname = fieldname || 'member';
        return this.reverse_node_table[node['@id']][fieldname];
    }

    get_nodes() {
        return _.values(this.node_table);
    }

    get_header() {
        return this.node_table['#__header__'];
    }

    reverse_node(node) {
        return this.reverse_node_table[node['@id']];
    }

    get_ancestors(node, field_name) {
        field_name = field_name || 'member';

        var rev_ptr = this.reverse_node_table[node['@id']];
        let ancestors = [];
        while (true) {
            if(!_.has(rev_ptr, field_name)) {
                break;
            }
            // follow only one chain for now (hence the [0])
            let rev_ancestor = rev_ptr[field_name][0];
            let rev_ancestor_id = rev_ancestor['@id'];
            // detect a circular reference
            if(_.includes(ancestors, rev_ancestor_id)) {
                break;
            }
            ancestors.push(rev_ancestor_id);
            rev_ptr = rev_ancestor;
        }
        return _.map(ancestors, x=>this.node_table[x]);
    }
}

function get_node_tree(node, fieldname) {
    if (!node) {
        return [];
    }
    fieldname = fieldname || 'member';
    return _.flatMap(_.castArray(node), n => {
        if (node_has_type(n, 'Structure') || !_.has(n, fieldname)) {
            return n;
        }
        var member_results = get_node_tree(n[fieldname], fieldname);
        member_results.push(n);
        return member_results;
    });
}

function node_has_type(node, type_to_match) {
    let type_to_match_list = _.castArray(type_to_match);
    if (!_.has(node, '@type')) {
        return false;
    }
    let node_type = _.castArray(node['@type']);

    for (let t of type_to_match_list) {
        if (_.includes(node_type, t)) {
            return true;
        }
    }
    return false;
}
    
module.exports = {
    flatten_doc, 
    index_doc, 
    flatten_node, 
    new_node_id, 
    dereference, 
    compose_field_values,
    merge_nodes,
    merge_indexed_docs,
    clone_node,
    get_node_tree,
    node_has_type,
    HAWG
};
